### PR TITLE
:fireworks:  Spec.json linter

### LINKED ISSUE #5937: Lint check that all spec fields have a `title` and `description` field set
## Tell us about the problem you're trying to solve
`title` and `description` are helpful in making the spec more human-readable in the UI. See the difference:

```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "additionalProperties": true,
  "properties": {
    "user_name": {
      "type": "string"
    }
  }
}
```
produces the following: 

Screen Shot 2021-09-08 at 9 04 23 PM

whereas this
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "user_name": {
      "type": "string",
      "description": "The username you use to login to the database",
      "title": "Username"
    }
  }
}
```
produces the following
Screen Shot 2021-09-08 at 9 05 16 PM

## Describe the solution you’d like
~1. Add an assertion in SAT's spec test that verifies that these meta fields are set on all fields in the spec. If the assertion fails, we should give a very clear error message to the user that helps them understand what's going on. 
2. Add a flag to the spec test config which allows skipping `tolerate_missing_metafields`. If this flag is set to true, skip this test. If the flag is not set or set to false, run the assertion
3. Add the flag to every existing connector's `acceptance-test-config.yaml` file (to avoid failing the build on all current connectors)~

It's probably better to have a custom flake/blake rule for this. It would be best if we can add them to the pre-commit hooks installed in https://github.com/airbytehq/airbyte/pull/4005

### PR DESCRIPTION
**Pull request by yaroslav-dudar**
## What
Implemented linter script which parse spec file #5937 

Linter applies for commited files only, so already existing connectors should not be affected

## How to execute
1. `pip install pre-commit`
2. `pre-commit run spec-linter -a`

## How
1. read spec file and serialize it as python dict object
2. get properties field from spec object
3. check if all fields from `FIELDS_TO_CHECK` exist in each property
4. if field has oneOf attribute - fetch all subobjects and for each of them goto step (2)

## Recommended reading order
1. `.pre-commit-config.yaml`
2. `tools/bin/spec_linter.py`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by midavadim — COMMENTED** [review]
According to the issue comment, we should check(and update) existing spec files. Has it been done already?

**Comment by yaroslav-dudar** [comment]
> According to the issue comment, we should check(and update) existing spec files. Has it been done already?

at this moment script performs check only. I can't find requirement about update

**Review by annalvova05 — APPROVED** [review]

File: tools/bin/spec_linter.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+# configure logging
+logging.basicConfig(format="%(message)s")
+
+
+def read_spec_file(spec_path: str):
+    with open(spec_path) as json_file:
+        try:
+            root_schema = json.load(json_file)["connectionSpecification"]["properties"]
+        except KeyError:
+            logging.error(f"\033[1m{spec_path}\033[0m: Couldn't find properties in connector spec.json")
```
Comment (antixar) [inline tools/bin/spec_linter.py:37]: Move this log printing to a separated function. It has same format.

File: tools/bin/spec_linter.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    schema: Mapping[str, Any],
+    parent_fields: List[str] = None,
+):
+    parent_fields = parent_fields if parent_fields else []
+    for field_name, field_schema in schema.items():
+        if not FIELDS_TO_CHECK.issubset(field_schema):
+            full_field_name = get_full_field_name(field_name, parent_fields)
+            logging.error(f"\033[1m{spec_path}\033[0m: Check failed for field \x1b[31;1m{full_field_name}\033[0m")
+            ERRORS[spec_path] += 1
+
```
Comment (antixar) [inline tools/bin/spec_linter.py:55]: I propose to use an aggregative approach: this logic has to validates all spec files and will print summary errors list if needed. For this every checking function should return an error or None. Using of the global value is not unclear

File: tools/bin/spec_linter.py (hunk 1 above)
Comment (antixar) [inline tools/bin/spec_linter.py:37]: Add a JSON exception too. A source JSON file can be broken

File: tools/bin/spec_linter.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            validate_schema(spec_path, oneof_schema["properties"], parent_fields + [field_name, str(index)])
+
+
+def fetch_oneof_schemas(schema: Mapping[str, Any]) -> List[Mapping[str, Any]]:
+    """
+    Finds sub-objects in oneOf field
+    """
+    if schema.get("oneOf") and schema["type"] == "object":
+        return [spec for spec in schema["oneOf"] if spec.get("properties")]
+    return []
```
Comment (antixar) [inline tools/bin/spec_linter.py:66]: I think all items of every "oneOf" have to be objects.  Can we return some error about unsupported types?

**Review by antixar — CHANGES_REQUESTED** [review]
Please add unit test set for local running

**Review by antixar — APPROVED** [review]
Great!

File: tools/git_hooks/spec_linter.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    return [spec for spec in schema.get("oneOf", []) if spec.get("properties")]
+
+
+def validate_field(
+    field_name: str,
+    schema: Mapping[str, Any],
+    parent_fields: Optional[List[str]] = None,
+) -> List[Tuple[str, str]]:
+    """
+    Validates single field objects and return errors if thay are exist
```
Comment (Phlair) [inline tools/git_hooks/spec_linter.py:118]:
```suggestion
    Validates single field objects and return errors if they exist
```

**Review by Phlair — APPROVED** [review]
This is great!

Would be useful to have some brief docstrings on the non-trivial functions.

I don't think we have any nested `oneOf`s at the moment but if we wanted to handle that might want to look at making validate_schema recursive. Non-blocking though.

**Comment by yaroslav-dudar** [comment]
> This is great!
> 
> Would be useful to have some brief docstrings on the non-trivial functions.
> 
> I don't think we have any nested `oneOf`s at the moment but if we wanted to handle that might want to look at making validate_schema recursive. Non-blocking though.

If I understand you properly current implementation should handle nested oneofs. I added this test case https://github.com/airbytehq/airbyte/blob/yaroslav-dudar/5937-spec-linter/tools/git_hooks/tests/test_spec_linter.py#L141

**Comment by Phlair** [comment]
Awesome, lgtm! Thanks

**Review by Phlair — APPROVED** [review]
Awesome, lgtm! Thanks

**Pull request merged**